### ISSUE TITLE
[Proton/Linux] Infinite loading on startup

### ISSUE THREAD

**Issue by Mershl** [Aragas/Bannerlord.MBOptionScreen#108]
Hi @Aragas,

Just tried Bannerlord on Linux using Proton 6.3-5. Performance is good and I got Harmony, ButterLib and UIExtenderEx running just fine. Once I enable Mod Configuration Menu (tested with v4.3.13.276411) the game is no longer starting up. It's stuck on the following screen, with the horse infinitly animated. Do you have a guess what could cause this?

![image](https://user-images.githubusercontent.com/520251/127753874-2df8f2b4-4ccd-4bda-8841-2606d41199d3.png)

Tested with:
![image](https://user-images.githubusercontent.com/520251/127753895-d13e8a3a-1450-48ac-afdc-99a1afa5dc79.png)

**Comment by Aragas** [comment]
Check this out for making the game work with Proton!
https://www.nexusmods.com/mountandblade2bannerlord/mods/612/?tab=bugs&issue_id=511684

**Comment by Mershl** [comment]
Hi @Aragas,
Thanks for the link. It shows the same behaviour as the last post of the linked bug report. Installing dotnet472 has not fixed it in my tests.

![image](https://user-images.githubusercontent.com/520251/129083692-2c96d72e-d299-4203-a0a0-42bf46ed611d.png)

Tested with protontricks 1.6.0 and Proton-GE-6.14. Can you confirm it should be running with dotnet472 with these versions? If yes, I'll then dive deeper to analyze the issue.

**Comment by Aragas** [comment]
I do remember that I got it to work on Manjaro by installing dotnet472.
It was some time ago, maybe there is some new breaking bug introduced since then, it will need some investigation

**Comment by AlpacaRotorvator** [comment]
I can reproduce this on the 1.6.4 branch. Installing dotnet472 actually makes the game get to the main menu instead of loading indefinitely, however it seems to completely bork older save files and any attempt at starting a new game results in a crash.

That crash seems to have nothing to do with MCM in particular btw. Running the game without any mods had the same result for me. It seems something in the actual game code that changed between the 1.5.x versions and 1.6.4 is causing this.

Some information that might help to reproduce the issue:
1- Per https://www.protondb.com/app/261550#18ZJ9In2-g , install dotnet472 on a fresh Proton 4.11 prefix
2- Change to a newer Proton version in order to get the launcher to work (I tried with both 6.3-7 and 6.21-GE-2)
3- Launch game and try to start a new save.

I might try and find out which actual game version broke this when I have some time on my hands. Since that protonDB report is from ~2 months ago, I suspect 1.6.1 or 1.6.2 perhaps would still work.

When running on a prefix without dotnet472, this is the load order I'm using:
* Harmony v2.1.1.65
* ButterLib v1.0.30.291416
* UIExtenderEx v2.1.8.291416
* MCM v4.3.14.291416

**Comment by AlpacaRotorvator** [comment]
Not sure of how much help this is, but I tried to go after what exactly happens when the game hangs, this is the exception I get on the console:

```
Unhandled Exception:
System.TypeLoadException: Could not load type of field 'MCM.UI.Functionality.DefaultIngameMenuScreenHandler:DataSource' (2) due to: Could not load file or assembly 'TaleWorlds.MountAndBlade.GauntletUI, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null' or one of its dependencies.
  at (wrapper managed-to-native) System.RuntimeTypeHandle.type_is_assignable_from(System.Type,System.Type)
  at System.RuntimeTypeHandle.CanCastTo (System.RuntimeType type, System.RuntimeType target) [0x00000] in <bee01833bcad4475a5c84b3c3d7e0cd6>:0 
  at System.RuntimeType.IsAssignableFrom (System.Type c) [0x00020] in <bee01833bcad4475a5c84b3c3d7e0cd6>:0 
  at TaleWorlds.MountAndBlade.Module.CollectModuleAssemblyTypes (System.Reflection.Assembly moduleAssembly) [0x00160] in <af016d4b02ca496ab6030f4bf122eb24>:0 
  at TaleWorlds.MountAndBlade.Module.AddSubModule (System.Reflection.Assembly subModuleAssembly, System.String name) [0x00015] in <af016d4b02ca496ab6030f4bf122eb24>:0 
  at TaleWorlds.MountAndBlade.Module.LoadSubModules () [0x0016a] in <af016d4b02ca496ab6030f4bf122eb24>:0 
  at TaleWorlds.MountAndBlade.Module.Initialize () [0x000a0] in <af016d4b02ca496ab6030f4bf122eb24>:0 
  at ManagedCallbacks.CoreCallbacksGenerated.Module_Initialize (System.Int32 thisPointer) [0x00006] in <3ff6a925356e4c6dacda1b08335aa771>:0 
  at (wrapper native-to-managed) ManagedCallbacks.CoreCallbacksGenerated.Module_Initialize(int)
[ERROR] FATAL UNHANDLED EXCEPTION: System.TypeLoadException: Could not load type of field 'MCM.UI.Functionality.DefaultIngameMenuScreenHandler:DataSource' (2) due to: Could not load file or assembly 'TaleWorlds.MountAndBlade.GauntletUI, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null' or one of its dependencies.
  at (wrapper managed-to-native) System.RuntimeTypeHandle.type_is_assignable_from(System.Type,System.Type)
  at System.RuntimeTypeHandle.CanCastTo (System.RuntimeType type, System.RuntimeType target) [0x00000] in <bee01833bcad4475a5c84b3c3d7e0cd6>:0 
  at System.RuntimeType.IsAssignableFrom (System.Type c) [0x00020] in <bee01833bcad4475a5c84b3c3d7e0cd6>:0 
  at TaleWorlds.MountAndBlade.Module.CollectModuleAssemblyTypes (System.Reflection.Assembly moduleAssembly) [0x00160] in <af016d4b02ca496ab6030f4bf122eb24>:0 
  at TaleWorlds.MountAndBlade.Module.AddSubModule (System.Reflection.Assembly subModuleAssembly, System.String name) [0x00015] in <af016d4b02ca496ab6030f4bf122eb24>:0 
  at TaleWorlds.MountAndBlade.Module.LoadSubModules () [0x0016a] in <af016d4b02ca496ab6030f4bf122eb24>:0 
  at TaleWorlds.MountAndBlade.Module.Initialize () [0x000a0] in <af016d4b02ca496ab6030f4bf122eb24>:0 
  at ManagedCallbacks.CoreCallbacksGenerated.Module_Initialize (System.Int32 thisPointer) [0x00006] in <3ff6a925356e4c6dacda1b08335aa771>:0 
```
I know zilch about C#, but it seems to be related to some reflection wizardry Butterlib's AccessTools2 use, around here?

https://github.com/Aragas/Bannerlord.MBOptionScreen/blob/f977016b8ccfa1241decfa2211427a9995d91700/src/MCM.UI/Functionality/DefaultIngameMenuScreenHandler.cs#L34-L35

ps.: sorry for flooding the thread, but I believe this is unrelated enough to my previous experiment to deserve a separate comment

**Comment by Aragas** [comment]
Thanks for the experiments with linux! Good to know it's now possible to do the trick with regular proton!

I can confirm the crash is a game API renaming issue. I'll fix that in the next release! Thanks for reporting the stacktrace!

**Comment by Aragas** [comment]
Actually, while there is a naming issue, the `dll not found` issue is not related to the issue. MCM was build with expected null handling here, so the missing field is not the issue. MCM will work fine in it's curent state still.

We had reports of the issue with Windows users too, so it's not a Linux\mono specific issues, but it's not consistent or common enough, so expect that the issue is due to misinstallation or some legacy files being present. Could you try to do a full clean install and confirm that the issue is still present?

**Comment by dsadaog** [comment]
I still get the same error as @AlpacaRotorvator , clean install and tried with Proton-6-31 / Proton-6.22-GE-2 / Proton-7.0rc2-GE-1

**Comment by mbotezatu** [comment]
Hello @Aragas

Since version 4.3.16 I get this error:

`Error while getting types and loadingCould not load type of field 'MCM.UI.Patches.OptionsGauntletScreenPatch:_spriteCategorySaveLoads' (0) due to: Could not load file or assembly 'TaleWorlds.MountAndBlade.GauntletUI, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null' or one of its dependencies. `

Proton: 6.21-GE-2

**Comment by Vargrak** [comment]
On a fresh install I could not get Bannerlord to launch with MCM out of the box on Proton 6.3-8. I would submit a log but I'm not sure where it is.

However, I used steam tinker launcher with winetricks and then installed dotnet48 when winetricks came up. My game is now able to get to the main menu, start custom battles, and campaign. I will update if I have any crashes.

**Comment by AlpacaRotorvator** [comment]
Oh hey @Aragas , I'm terribly sorry for the delay in answering, and I wish you a happy new year while we're here!

So I did try a fresh install, deleted every single extant file from the BL folder and my proton prefix beforehand, downloaded the newest available versions of MCM and its dependencies and tried running on Proton 6.3-8, but the bug still persists. The exception I get this time is this:
```
Unhandled Exception:
System.TypeLoadException: Could not load type of field 'MCM.UI.Patches.OptionsGauntletScreenPatch:_spriteCategorySaveLoads' (0) due to: Could not load file or assembly 'TaleWorlds.MountAndBlade.GauntletUI, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null' or one of its dependencies.
  at (wrapper managed-to-native) System.RuntimeTypeHandle.type_is_assignable_from(System.Type,System.Type)
  at System.RuntimeTypeHandle.CanCastTo (System.RuntimeType type, System.RuntimeType target) [0x00000] in <2b59a54be42b4db59a5b1855eee0c040>:0 
  at System.RuntimeType.IsAssignableFrom (System.Type c) [0x00020] in <2b59a54be42b4db59a5b1855eee0c040>:0 
  at TaleWorlds.MountAndBlade.Module.CollectModuleAssemblyTypes (System.Reflection.Assembly moduleAssembly) [0x00160] in <af016d4b02ca496ab6030f4bf122eb24>:0 
  at TaleWorlds.MountAndBlade.Module.AddSubModule (System.Reflection.Assembly subModuleAssembly, System.String name) [0x00015] in <af016d4b02ca496ab6030f4bf122eb24>:0 
  at TaleWorlds.MountAndBlade.Module.LoadSubModules () [0x0016a] in <af016d4b02ca496ab6030f4bf122eb24>:0 
  at TaleWorlds.MountAndBlade.Module.Initialize () [0x000a0] in <af016d4b02ca496ab6030f4bf122eb24>:0 
  at ManagedCallbacks.CoreCallbacksGenerated.Module_Initialize (System.Int32 thisPointer) [0x00006] in <3ff6a925356e4c6dacda1b08335aa771>:0 
  at (wrapper native-to-managed) ManagedCallbacks.CoreCallbacksGenerated.Module_Initialize(int)
[ERROR] FATAL UNHANDLED EXCEPTION: System.TypeLoadException: Could not load type of field 'MCM.UI.Patches.OptionsGauntletScreenPatch:_spriteCategorySaveLoads' (0) due to: Could not load file or assembly 'TaleWorlds.MountAndBlade.GauntletUI, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null' or one of its dependencies.
  at (wrapper managed-to-native) System.RuntimeTypeHandle.type_is_assignable_from(System.Type,System.Type)
  at System.RuntimeTypeHandle.CanCastTo (System.RuntimeType type, System.RuntimeType target) [0x00000] in <2b59a54be42b4db59a5b1855eee0c040>:0 
  at System.RuntimeType.IsAssignableFrom (System.Type c) [0x00020] in <2b59a54be42b4db59a5b1855eee0c040>:0 
  at TaleWorlds.MountAndBlade.Module.CollectModuleAssemblyTypes (System.Reflection.Assembly moduleAssembly) [0x00160] in <af016d4b02ca496ab6030f4bf122eb24>:0 
  at TaleWorlds.MountAndBlade.Module.AddSubModule (System.Reflection.Assembly subModuleAssembly, System.String name) [0x00015] in <af016d4b02ca496ab6030f4bf122eb24>:0 
  at TaleWorlds.MountAndBlade.Module.LoadSubModules () [0x0016a] in <af016d4b02ca496ab6030f4bf122eb24>:0 
  at TaleWorlds.MountAndBlade.Module.Initialize () [0x000a0] in <af016d4b02ca496ab6030f4bf122eb24>:0 
  at ManagedCallbacks.CoreCallbacksGenerated.Module_Initialize (System.Int32 thisPointer) [0x00006] in <3ff6a925356e4c6dacda1b08335aa771>:0 
```

**Comment by Aragas** [comment]
This is a weird bug. I'll try to setup manjaro again and check the issue, for some reason the runtime can't find a library it should.

**Comment by Aragas** [comment]
I at least know that this issue should be at least a year old. We had this bug reported, but I wasn't able to reproduce it

**Comment by ConnorPivovarov** [comment]
Hi Aragas, I was the Nexus-User who first reported Linux issues to you. I also keep encountering never ending load screens with this. I didn't report back for a while because of some real life stuff, but I hope we can find a way to fix this. I have a freshly set up Bannerlord on Manjaro and I am ready to do some tests for you.

**Comment by Aragas** [comment]
Forwarding this [message](https://forums.nexusmods.com/index.php?showtopic=8605808/#entry104400353)!

>  Regarding using mods when playing Bannerlord  while using a Linux OS:
> 
> I wasn't able to get any mods using Harmony to work. Bannerlord wouldn't get past the first loading screen, no CTD, just endlessly loading.
> 
> Until I finally found this post here on comments page 10 and voila, Bannerlord with mods running just fine...
> 
> --------------------------------------------------
> Hi. I mentioned an "infinite loop" bug when running this mod with proton on Linux before.
> 
> I managed to fix it, and I want to share it with all other Linux users.
> 
> So the culprit was that "dotnet472" is broken in the latest builds of WINE (specifically after 5.18 or something).
> To make it work, we need a WINE that can successfully install it.
> 
> The setup is the following:
> 1. In steam, go to properties go to "properties" (of Bannerlord) ->
> "compatibility." Then tick "Force the use of a specific SteamPlay
> compatibility tool." Select "Proton 5.0-10" (you must pick this exact version. It is the latest version where the aforementioned "dotnet472" works, as far as I know).
> 2. Go to your console and type "protontricks 261550 dotnet472." The number
> 251550 is the ID of Bannerlord. Note that if you don't have
> "protontricks," go search on google and find out how to install it in
> your own distribution.
> 3. ????
> 4. PROFIT xd
> --------------------------------------------------
> 
> I have to add, that after installing dotnet472 I was able to change the forced compatibility tool all the way up back to "proton experimental" to play Bannerlord with all my mods. Installing dotnet472 seems to do the trick regardless of the proton version.
> 
>

**Comment by Aragas** [comment]
Okay, v4.3.19 might fix the issue, can someone confirm?

**Comment by kol-93** [comment]
@Aragas 

I've tried many preconditions (different proton versions, clean wine roots, different versions of mod, etc).
But this shit really works - added symlinks to those DLLs at `Modules/Bannerlord.MBOptionScreen/bin/Win64_Shipping_Client`:

 - `TaleWorlds.MountAndBlade.GauntletUI.dll` -> `../../../Native/bin/Win64_Shipping_Client/TaleWorlds.MountAndBlade.GauntletUI.dll`
 - `TaleWorlds.MountAndBlade.GauntletUI.AutoGenerated.dll` -> `../../../Native/bin/Win64_Shipping_Client/TaleWorlds.MountAndBlade.GauntletUI.AutoGenerated.dll`
 - `TaleWorlds.MountAndBlade.Platform.PC.dll` -> `../../../Native/bin/Win64_Shipping_Client/TaleWorlds.MountAndBlade.Platform.PC.dll`
 - `TaleWorlds.MountAndBlade.View.dll` -> `../../../Native/bin/Win64_Shipping_Client/TaleWorlds.MountAndBlade.View.dll`
 - `SandBox.View.dll` -> `../../../SandBox/bin/Win64_Shipping_Client/SandBox.View.dll`
 - `SandBox.GauntletUI.dll` -> `../../../SandBox/bin/Win64_Shipping_Client/SandBox.GauntletUI.dll`
 - `SandBox.GauntletUI.AutoGenerated.dll` -> `../../../SandBox/bin/Win64_Shipping_Client/SandBox.GauntletUI.AutoGenerated.dll`
 - `SandBox.ViewModelCollection.dll` -> `../../../SandBox/bin/Win64_Shipping_Client/SandBox.ViewModelCollection.dll`
 - `SandBox.dll` -> `../../../SandBox/bin/Win64_Shipping_Client/SandBox.dll`
 - `StoryMode.dll` -> `../../../StoryMode/bin/Win64_Shipping_Client/StoryMode.dll`
 - `StoryMode.ViewModelCollection.dll` -> `../../../StoryMode/bin/Win64_Shipping_Client/StoryMode.ViewModelCollection.dll`
 - `StoryMode.View.dll` -> `../../../StoryMode/bin/Win64_Shipping_Client/StoryMode.View.dll`
 - `StoryMode.GauntletUI.dll` -> `../../../StoryMode/bin/Win64_Shipping_Client/StoryMode.GauntletUI.dll`
 - `StoryMode.GauntletUI.AutoGenerated.dll` -> `../../../StoryMode/bin/Win64_Shipping_Client/StoryMode.GauntletUI.AutoGenerated.dll`

I really don't want to know how .NET Assembly Loading mechanism works and why is it so weird, but I hope this helps you to found any solution.

**Comment by ConnorPivovarov** [comment]
> @Aragas
> 
> I've tried many preconditions (different proton versions, clean wine roots, different versions of mod, etc). But this shit really works - added symlinks to those DLLs at `Modules/Bannerlord.MBOptionScreen/bin/Win64_Shipping_Client`:
> 
>     * `TaleWorlds.MountAndBlade.GauntletUI.dll` -> `../../../Native/bin/Win64_Shipping_Client/TaleWorlds.MountAndBlade.GauntletUI.dll`
> 
>     * `TaleWorlds.MountAndBlade.GauntletUI.AutoGenerated.dll` -> `../../../Native/bin/Win64_Shipping_Client/TaleWorlds.MountAndBlade.GauntletUI.AutoGenerated.dll`
> 
>     * `TaleWorlds.MountAndBlade.Platform.PC.dll` -> `../../../Native/bin/Win64_Shipping_Client/TaleWorlds.MountAndBlade.Platform.PC.dll`
> 
>     * `TaleWorlds.MountAndBlade.View.dll` -> `../../../Native/bin/Win64_Shipping_Client/TaleWorlds.MountAndBlade.View.dll`
> 
>     * `SandBox.View.dll` -> `../../../SandBox/bin/Win64_Shipping_Client/SandBox.View.dll`
> 
>     * `SandBox.GauntletUI.dll` -> `../../../SandBox/bin/Win64_Shipping_Client/SandBox.GauntletUI.dll`
> 
>     * `SandBox.GauntletUI.AutoGenerated.dll` -> `../../../SandBox/bin/Win64_Shipping_Client/SandBox.GauntletUI.AutoGenerated.dll`
> 
>     * `SandBox.ViewModelCollection.dll` -> `../../../SandBox/bin/Win64_Shipping_Client/SandBox.ViewModelCollection.dll`
> 
>     * `SandBox.dll` -> `../../../SandBox/bin/Win64_Shipping_Client/SandBox.dll`
> 
>     * `StoryMode.dll` -> `../../../StoryMode/bin/Win64_Shipping_Client/StoryMode.dll`
> 
>     * `StoryMode.ViewModelCollection.dll` -> `../../../StoryMode/bin/Win64_Shipping_Client/StoryMode.ViewModelCollection.dll`
> 
>     * `StoryMode.View.dll` -> `../../../StoryMode/bin/Win64_Shipping_Client/StoryMode.View.dll`
> 
>     * `StoryMode.GauntletUI.dll` -> `../../../StoryMode/bin/Win64_Shipping_Client/StoryMode.GauntletUI.dll`
> 
>     * `StoryMode.GauntletUI.AutoGenerated.dll` -> `../../../StoryMode/bin/Win64_Shipping_Client/StoryMode.GauntletUI.AutoGenerated.dll`
> 
> 
> I really don't want to know how .NET Assembly Loading mechanism works and why is it so weird, but I hope this helps you to found any solution.

can confirm this also fixes issues on my end. I think UIExtender has the same issues. So if anyone were to run UIExtenter without MBOptions, then they'd need to add these symlinks inside the UIExtender module.

**Comment by nelsam** [comment]
Just my two denar to add to this:

First, for reference: I'm on gentoo linux, and the current version of Bannerlord is `1.7.2` (`1.8.0` is the  current beta).  I have mods working in `1.7.1`, `1.7.2`, and the `1.8.0` beta, although `1.7.1` keeps resetting my weapon switch key bindings so I can't really play it.

---

I was unable to get mods working using the symlinks above.  I believe I set them up correctly, but I didn't see any change in the loading.  Here's the command I used to get those DLLs symlinked (which I ran from my `[..]/Modules/Bannerlord.MBOptionScreen/bin/Win64_Shipping_Client` directory): `coremods=( Native Sandbox StoryMode ); for coremod in ${coremods[@]}; do find ../../../${coremod} -iname '*.dll' -exec ln -s {} . \; ; done`.  I also tried including the `CustomBattle` and `SandboxCore` modules, without any success.

---

I _was_ able to get mods working with dotnet472, but only following a _very specific_ sequence:

1. Install [Proton-5.5-GE-1](https://github.com/GloriousEggroll/proton-ge-custom/releases/tag/5.5-GE-1).
    - When I started with the built-in `proton 5.0-10`, dotnet472 installed correctly - but then Bannerlord crashed on launch when I switched back to a more recent version.  Using `5.5-GE-1` did not result in that crash.
2. Install [GE-Proton-7-20](https://github.com/GloriousEggroll/proton-ge-custom/releases/tag/GE-Proton7-20).
    - I honestly didn't try any other versions, and now that I have mods working, I am too scared of breaking it to try anything else.
3. Install `protontricks` version `1.7.0`.
    - gentoo's repositories had `1.8.2` at the time of this writing, but `1.8.2` failed when run against `Proton-5.5-GE-1`.  I had to fall back to `1.7.0` in order to get any success with older versions of proton.
4. Delete the whole compatdata directory. `rm -rf [..]/steamapps/compatdata/261550`
    - I was never able to get dotnet472 successfully installed to an existing prefix.  I had to delete the prefix and make sure it was created using an older version of proton.
    - If you have an existing game, back up your saves before deleting the prefix: `cp -r [..]/steamapps/compatdata/261550/pfx/drive_c/users/steamuser/Documents/Mount\ and\ Blade\ II\ Bannerlord/ /path/to/game/backup/dir`
5. In `Bannerlord`'s compatibility settings, force it to use `Proton-5.5-GE-1`
6. Start `Bannerlord` _in steam_ just to create the prefix.  Just close the launcher after it opens.
7. `protontricks 261550 dotnet472`
    - Expect several installers for several versions of dotnet.  Go through the installer each time.
    - When asked if I wanted to "Restart Windows", I said "yes".  I doubt it matters.
8. In `Bannerlord`'s compatibility settings, force it to use `GE-Proton-7-20`
9. Start `Bannerlord`, enable any mods you like.
    - When mods crash, the game will try to pop up a dialog box.  Depending on your window manager, it _may_ pop up _behind_ the game window - so if it looks like you're getting infinite loading, first check if there's a dialog box hidden behind the game.  It may just be that one of the mods crashed.

---

This was after a lot of failed attempts, after trying several versions of proton as well as switching back and forth between local steam and flatpak steam (which runs in a container).  In the end, it didn't seem to matter whether I used local or flatpak steam - but the versions of proton and protontricks were _very_ picky, on my system.

**Comment by extract** [comment]
Hi, I followed @nelsam's instructions and can confirm it works. However, this process is not smooth. I'm assuming this is due to this project having different targets of the .net framework. This is with Arch Linux so the steps seems necessary on more than just previously specified system.

Maybe someone could attempt to change .net framework target for Linux builds in this repo, or something, to match Bannerlord's version. Whatever that is.

**Comment by Aragas** [comment]
Fixed with v5.1.1
